File: problems/1679/1679.max-number-of-k-sum-pairs.py
#
# @lc app=leetcode id=1679 lang=python3
#
# [1679] Max Number of K-Sum Pairs
#


# @lc code=start
from collections import defaultdict


class Solution:
    def maxOperations(self, nums: List[int], k: int) -> int:
        # O(N) time, O(N) space

        res = 0
        seen = defaultdict(int)

        for val in nums:
            if seen[val] > 0:
                res += 1
                seen[val] -= 1
            else:
                seen[k - val] += 1

        return res


# @lc code=end

File: problems/1679/1679.max-number-of-k-sum-pairs.go
/*
 * @lc app=leetcode id=1679 lang=golang
 *
 * [1679] Max Number of K-Sum Pairs
 */

package solution

// @lc code=start
func maxOperations(nums []int, k int) int {
	// O(N) time, O(N) space

	dict := make(map[int]int)
	res := 0

	for _, val := range nums {
		if cnt, ok := dict[val]; ok {
			res += 1
			if cnt > 1 {
				dict[val] -= 1
			} else {
				delete(dict, val)
			}
		} else {
			dict[k-val]++
		}
	}

	return res
}

// @lc code=end
